### Seq2Seq model for English to Chinese translation (baseline model)

In [2]:
import json
import numpy as np
import pickle
from collections import Counter
import string
import re
import torch
import torch.nn as nn

### Choosing, loading, and cleaning dataset

- going to use datasets from: https://www.kaggle.com/datasets/qianhuan/translation?resource=download

In [3]:
train_set_path = "dataset/translation2019zh_train.json"

train_set = []
with open(train_set_path) as f:
    for line in f:
        train_set.append(json.loads(line))

print(len(train_set))
print(train_set[0])

5161434
{'english': 'For greater sharpness, but with a slight increase in graininess, you can use a 1:1 dilution of this developer.', 'chinese': '为了更好的锐度，但是附带的会多一些颗粒度，可以使用这个显影剂的1：1稀释液。'}


We want to lower the size of this dataset, for testing purposes.
- right now 5.1M sentences
- let's make it 500,000 sentences...

In [4]:
# get 10,000 random indices 
sampled_indices = np.random.choice(len(train_set), 500000)

train_subset = [train_set[i] for i in sampled_indices]
print(train_subset[0])
with open('dataset/train_set_mini.pkl', 'wb') as f:
    pickle.dump(train_subset, f)

{'english': 'Exhaustion combined with overindulgence could result in minor health problems.', 'chinese': '过度放纵和疲劳会带来身体微恙。'}


In [5]:
with open('dataset/train_set_mini.pkl', 'rb') as f:
    train_set_mini = pickle.load(f)

print(train_set_mini[0])

{'english': 'Exhaustion combined with overindulgence could result in minor health problems.', 'chinese': '过度放纵和疲劳会带来身体微恙。'}


Train set made. 
Now to work on the actual architecture

## Preprocessing steps:
- we want to maintain vocabulary for english and chinese. 
    - this is simple, just use a counter() and then limit it to if it appears within 5

- In terms of encoding, we want to use sequence input, so a sentence becomes a list [3, 100, 8, 9], where each number corresponds to the index of the word in the dictionary
    - we want to do this because LSTMs keep track of the word relationships at the sentence level
    - then use nn.Embedding?
        - nn.Embedding allows us to create a matrix representing the vocabulary. It allows us to train some nuance into the words, where instead of typical BoW where each word is just an index, each word is now a vector, which allows us to train some meaning into the word
        - the embedding is a matrix size (vocab length, dim). Length vocab length because each row in the matrix corresponds to a word in the vocab, ( row of index = index of word in vocab :) )
        

# now to work on the vocabulary

In [6]:
## helper functions 
def remove_punctuation(text):
    '''
    Get rid of all punctuation from string text
    '''
    return text.translate(str.maketrans('', '', string.punctuation))

def get_words_from_sentence(s):
    '''
    Gets words from sentence 
    '''
    return s.split(' ')

def clean_en_pair(pair):
    '''
    Cleans the english from the pair 
    '''
    return get_words_from_sentence(remove_punctuation(pair['english']).lower())



In [ ]:
def get_en_vocab(train_set):
    '''
    get_en_dict:
        Gets an english vocab from train_set as a dict 
    '''
    # get only the english sentences, list of strings 
    en_sentences = [clean_en_pair(pair) for pair in train_set]
    en_sentences_flattened = [word for sentence in en_sentences for word in sentence]
    print(f"Words pre-clean {len(en_sentences_flattened)}")
    en_sentences_flattened = [word for word in en_sentences_flattened if word != '']
    print(f"Words post-clean {len(en_sentences_flattened)}")
    
    word_counts = Counter(en_sentences_flattened)
    # with word counts, now we limit the vocabulary to words that happen at least 5 times
    en_vocab = {}
    # {word: index}
    idx = 0
    for word in ["<SOS>", "<EOS>", "<UNK>"]:
        en_vocab[word] = idx 
        idx += 1
    for word, occurrences in word_counts.items():
        if occurrences >= 5:
            en_vocab[word] = idx 
            idx += 1
    return en_vocab

def remove_zh_punctuation(text):
    cleaned = re.sub(r'[，。！？【】（）《》“”‘’、]', '', text)
    cleaned = re.sub(r'\s+', '', cleaned)
    return cleaned

def get_zh_vocab(train_set):
    '''
    get_zh_vocab:
        Gets an zh vocab from train_set as a dict 
    '''
    zh_sentences = [list(remove_zh_punctuation(pair['chinese'])) for pair in train_set]
    zh_sentences_flattened = [word for sentence in zh_sentences for word in sentence]
    print(len(zh_sentences_flattened))

    word_counts = Counter(zh_sentences_flattened)
    zh_vocab = {}

    idx = 0 
    for word in ["<SOS>", "<EOS>", "<UNK>"]:
        zh_vocab[word] = idx 
        idx += 1 
    for word, occurrences in word_counts.items():
        if occurrences >= 2: 
            zh_vocab[word] = idx 
            idx += 1 
    return zh_vocab

en_vocab = get_en_vocab(train_set_mini)
print(len(en_vocab))

zh_vocab = get_zh_vocab(train_set_mini)
print(len(zh_vocab))

Words pre-clean 9586368
Words post-clean 9537657
57280
16402788
6403


In [8]:
with open('vocab/en_vocab.pkl', 'wb') as f:
    pickle.dump(en_vocab, f)

with open('vocab/zh_vocab.pkl', 'wb') as f:
    pickle.dump(zh_vocab, f)

In [9]:
with open('vocab/en_vocab.pkl', 'rb') as f:
    en_vocab = pickle.load(f)

with open('vocab/zh_vocab.pkl', 'rb') as f:
    zh_vocab = pickle.load(f)

### Model architecture building
- 2 LSTM's are the backbone
- also build a higher level Seq2Seq model as abstraction of the entire model 
- nn.Embedding() as a variable for both Encoder and Decoder 
    - use vocab_size as row length, by the embedding dim as the column length
- Encoder will be english, decoder will be chinese 

### nn.LSTM

- sequence models are central to NLPl they are models where there is some sort of dependence through teime between inputs. 
- a recurrent neural network is a network that maintains some kind of state.
- for example its output could be used as part of the next input, so that information c an propagate along as the network passes over the sequence.
- In the case of an LSTM, for each element in the sequence, there is a corresponding hidden state ht, which in principle contains infromation from arbitrary points earlier in the sequence. 
- we can use the hidden state to predict words in a language model, pos, and a myriad of oether things.

LSTMs in pytorch:
- pytorch LSTM expects all of its inputs to be 3D tensors. The semantics of the axes of these tensors is important. The first axis is the sequence itself, the second indixes instances in the mini batch, and the third indexes elements of the input. 
- ignore mini batching, we will always just have 1 dimension on th second axis.
- If we want to run the sequence model over the sentence "The cow jumped" our input should look like:

[
    q (the)
    q (cow)
    q (jumped)
]
Except remember there is an additional 2nd dimension with size 1, (this dimension )

Initializing an LSTM:
```python
lstm = nn.LSTM(3, 3) #input dim is 3, output dim is 3 
```

input_size = 3: this means each input vector at a time step is of length 3. All inputs must have 3 columns, (n x 3). 
- each sequence = a list of input vectors (one per timestemp)
- each input vector = size input_size 

input_size = 3, then your input tensor shape for 1 batch would be:
    (seq_len, batch_size, 3)

- What does this mean for the embedding layer?
    - it must also be dimension (vocab_size, 3), since each token is mapped to a vecotr using the embedding. This embedding becomes the input at each timestep for the LSTM, and the LSTM accepts vector of dimensions (3).

- using nn.LSTM example:
    
```python
self.lstm = nn.LSTM(embedding_dim, hidden_dim)  #(embedding_dim) is the dimension of the embedding dim, and then hidden_dim is also essentially a hyperparameter, it's the dimension of the hidden state 
```

**For most LSTM applications we will need a linear layer to learn the mapping from hidden state to the tag space. but in the case of the encoder, you don't need it! Since the linear layer is essentially the classifier layer, that learns to 
interpret the hidden layer.**


### forward:
- in forward we will have
```
lstm_out, _ = self.LSTM(embeds.view(len(sentence), 1, -1))
```
- what exactly is going on here? Well, basically embeds is a tensor of dimensions (number of tokens in sentence, embedding dim) (**since rememember that the embedding layer takes each word index, and pulls up the corresponding row from the vocabulary)
- we want to reshape it to (sequence length, batch_size, input_size)
    - reshaping is: sequence length( the length of the sentence )
    - batch size is 1, since its one sentence at a time 
    - input_size = -1: it assumes that the length will be embedding_dim 
- **embeds.view**: is a way to reshape in LSTM

- forward in the encoder should only return the hidden state and cell state, since its what matters 


## backpropagation
- pytorch under the hood tracks operations on tensors with require_grad = True. All nn.modules like nn.Linear and nn.LSTM already register their parameters with requires_grad=True, so as long as its connected correctly in the forward pass, pytorch will handle the gradients during backprop. 

In [10]:
## Encoder: English layer 

class Encoder(nn.Module):
    def __init__(self, embedding_dim, vocab_size, hidden_dim):
        super(Encoder, self).__init__()
        self.embeddings = nn.Embedding(vocab_size, embedding_dim)
        self.LSTM = nn.LSTM(embedding_dim, hidden_dim) # initialize an LSTM, with embedding_dim, and hidden_dim hyperparameters 
    
    def forward(self, sentence):
        embeds = self.embeddings(sentence)  # remember that sentence has to the in [word_index0, word_index1, word_index2] form
        out , (h_n, c_n) = self.LSTM(embeds.view(len(sentence), 1, -1)) # one timestep at a tiem 
        return out, (h_n, c_n)
        

In [11]:
## Test example pass through the encoder 
encoder = Encoder(embedding_dim=3, vocab_size= len(en_vocab), hidden_dim=5)
# now remember that for forward we pass a sentence as the list of words mapped to the indices they show up in the vocab, such as [45, 18, 28]
sentence = "I love bread."

input_words = get_words_from_sentence(remove_punctuation(sentence).lower())
# now map the inputs to the vocab 
input_indices = [en_vocab[word] for word in input_words] 
# now that I think about it, we probably want a function that does this, so that we don't get hit with a KeyError and actually use our <unk> token lul
input_indices

[78, 293, 6631]

In [12]:
# with the input_indices, we can now throw it through the encoder?
input_indices_tensor = torch.tensor(input_indices, dtype=torch.long)
output, hidden = encoder.forward(input_indices_tensor)
hidden # this makes sense, we have both hidden and cell states :)

(tensor([[[-0.4061,  0.0178,  0.2490,  0.0442, -0.0523]]],
        grad_fn=<StackBackward0>),
 tensor([[[-0.7400,  0.0740,  0.5608,  0.1064, -0.1235]]],
        grad_fn=<StackBackward0>))

### Decoder
- so the decoder is another LSTM, taking as input "a large fixed-dimensional vector representation", and then use another LSTM to extract the output sequence from that vector 
- we can just pass in h_n and c_n in the decoder LSTM as parameter for this! 
- and then for forward we just run the linear layer and then run the log_softmax to get the logits?

- What about the Embedding layer?
    - we also need an embedding layer ( used in both training and inference )
    - training phase we have "ground-truth" tokens, we need the "ground-truth" tokens we need the embedding layer to make them tensors and to feed each token through the decoder

### Teacher forcing and backpropagation
- At a time step t, the input is the actual target sequence from t - 1 !. This makes sense, we give it the "correct" input from the time before, and have it try to predict the input now.  the t-1 token is called the "ground truth" token, is passed through the embedding layer specifically trained for the target language vocabulary
- the output is the predicted timestep t token, and what you use to compare as loss is the actual t token. 

**forward step in the decoder**
- when making a prediction, you either use the correct previous token (teacher forcing during training), or you use the previous prediction (inference)
- during prediction we will use nn.LSTM. You want to pass the token at t-1's tensor, AND the previous c_n and h_n from the decoder! it's the recurrent aspect of the RNN

- we also do feed in the encoder h_n, and c_n, in the first step :).
- don't need to worry about the rest, because it already does it implicitly by nn.LSTM

- also to enforce teacher forcing, we have to do one time step at a time, instead of all at once 
    - so instead, we go through every single input one at time 

- in the no teacher YOU USE THE PREVIOUS PREDICTION OF THE TIME(after it is )

In [13]:
## Configurations 
MAX_RESPONSE_LENGTH = 10

In [14]:
### Luong attention general form (with Linear layer)
# unsure what the shape should be at the moment 
# make just a linear transformation with no bias
embedding_dim = 32 
hidden_dim = 128 

general_attn_layer = nn.Linear(in_features=hidden_dim, out_features=hidden_dim, bias=False)

class LuongAttnDecoder(nn.Module):
    def __init__(self, embedding_dim, vocab_size, hidden_dim, device):
        super(LuongAttnDecoder, self).__init__()
        self.embeddings = nn.Embedding(vocab_size, embedding_dim)
        self.LSTM = nn.LSTM(embedding_dim, hidden_dim)
        self.linear = nn.Linear(hidden_dim * 2, vocab_size) # 2x hidden dim now for result after concat attention input
        self.device = device 
    
    def word_to_tensor(self, word):
        '''
        takes a single wrod and gets the corresponding tensor
        '''
        word_lst = get_words_from_sentence(remove_zh_punctuation(word))
        indices = [zh_vocab[word] for word in word_lst]
        # get tensor 
        return torch.tensor(indices, dtype=torch.long).to(self.device)

    def forward(self, hidden, encoder_out, sentence=None):
        '''
        does the forward propagation. If sentence is provided, then we do teacher-forcing. Else we assume it is inference  
            Params:
                hidden: the hidden state passed from the previous 
                sentence: a sentence to be used for teacher-forcing, as a tensor 
                Make sure the teacher-forcing sentence is sliced to not include the last token [:-1]
        '''
        # teacher-forcing training, we are going to "undo" vectorization]
        all_outputs = []
        if sentence is not None:
            embeds_tensor = self.embeddings(sentence)
            for word_tensor in embeds_tensor:
                out, hidden = self.LSTM(word_tensor.view(1, 1, -1), hidden)
                # pass encoder out to attention layer 
                attn_scores = general_attn_layer(encoder_out) @ hidden[0].squeeze()
                # now with attn scores, we want to softmax the scores 
                softmaxed_scores = torch.nn.functional.softmax(attn_scores, dim=0)
                # multiply by encoder_out
                # now that they are softmaxed, we want to multiply by all encoder states to give a weighted tensor, we can broadcast it as well 
                weighted_encoder_hidden_states = softmaxed_scores * encoder_out.squeeze()
                # sum the tensor 
                context = torch.sum(weighted_encoder_hidden_states, dim=0).view(1, 1, -1)
                # concat the context vector with the hidden state 
                combined_tensor = torch.concat([context, hidden[0]], dim=-1)
                logits = self.linear(combined_tensor)
                all_outputs.append(logits)
        else:
            start_token = self.word_to_tensor('<SOS>')
            # run through embedding layer
            prev_char = start_token
            for i in range(MAX_RESPONSE_LENGTH):
                if prev_char.item() == 1:
                    break
                embeds = self.embeddings(prev_char).to(self.device)
                out, hidden = self.LSTM(embeds.view(1, 1, -1), hidden)
                attn_scores = general_attn_layer(encoder_out) @ hidden[0].squeeze()
                softmaxed_scores = torch.nn.functional.softmax(attn_scores, dim=0)
                weighted_encoder_hidden_states = softmaxed_scores * encoder_out.squeeze()
                context = torch.sum(weighted_encoder_hidden_states, dim=0).view(1, 1, -1)
                combined_tensor = torch.concat([context, hidden[0]], dim=-1)
                logits = self.linear(combined_tensor)
                all_outputs.append(logits)
                pred_idx = torch.argmax(logits, dim=2).item()
                prev_char = torch.tensor(pred_idx, dtype=torch.long, device=self.device)
        return torch.cat(all_outputs, dim=0)

In [15]:
## functions to take a sentence and turn it into a tensor, adding <sos> and <eos>
def sequence_to_tensor_en(sequence):
    '''
    takes sequence and converts to tensor 
    '''
    # add "<SOS> and <EOS>"
    words = get_words_from_sentence("<SOS> " + remove_punctuation(sequence).lower() + " <EOS>")
    
    # convert to indices, reverting to <UNK> token
    word_indices = [ en_vocab[word] if word in en_vocab else en_vocab["<UNK>"] for word in words ]
    return torch.tensor(word_indices, dtype=torch.long)
    

def sequence_to_tensor_zh(sequence):
    '''
    takes sequence and converts to chinese tensor 
    '''
    words = (["<SOS>"] + list(remove_zh_punctuation(sequence)))
    words.append("<EOS>")
    
    word_indices = [ zh_vocab[word] if word in zh_vocab else zh_vocab["<UNK>"] for word in words ]
    return torch.tensor(word_indices, dtype=torch.long)

In [16]:
## new 100,000 len
print("Length of train ", len(train_set_mini))
print("Length of zh dictionary ", len(zh_vocab))
print("Length of english dictionary ", len(en_vocab))

Length of train  500000
Length of zh dictionary  6403
Length of english dictionary  57280


In [19]:
def zh_tensor_outputs_to_sentence(output_tensor):
    s = ''
    zh_vocab_lst = list(zh_vocab.keys())
    for word_tensor in output_tensor:
        pred_idx = torch.argmax(word_tensor, dim=-1).item()
        s += zh_vocab_lst[pred_idx]
    return s 

## Training loop

Does mps save time?
琥ä纷随哗良钮眼黎s
Number of trains 1000
Loss 6.893100890159607
time for 1000 : mps: 13.119836807250977

CPU:
Number of trains 1000
Loss 6.896958403587341
time for 1000 : cpu: 23.142292022705078

In [20]:
import time

def train(num_epochs, training_data, encoder, decoder, device, lr=0.001):
    optimizer = torch.optim.Adam(
    list(encoder.parameters()) + list(decoder.parameters()), lr=lr
)
    count = 0
    total_loss = 0
    # see random prediction
    predict_en_sequence = "I love bread" 
    predict_en_tensor = sequence_to_tensor_en(predict_en_sequence).to(device)
    out, predict_hidden = encoder(predict_en_tensor)
    out = out.to(device)
    print(zh_tensor_outputs_to_sentence(decoder.forward(predict_hidden, encoder_out=out)))
    start_time = time.time()
    for i in range(num_epochs):
        for pair in training_data:
            count += 1
            if count % 10000 == 0:
                print(f"Number of trains {count}")
                # print the loss 
                print(f"Loss {total_loss / 10000}")
                total_loss = 0
            english = pair['english']
            zh = pair['chinese']
            en_tensor = sequence_to_tensor_en(english)
            zh_tensor = sequence_to_tensor_zh(zh)
            # pass to device 
            en_tensor = sequence_to_tensor_en(english).to(device)
            zh_tensor = sequence_to_tensor_zh(zh).to(device)

            out, hidden = encoder.forward(en_tensor)
            target = zh_tensor[1:]
            predicted = decoder.forward(hidden, encoder_out=out, sentence = zh_tensor[:-1])
            loss = nn.functional.cross_entropy(torch.squeeze(predicted), target)
            total_loss += loss.item()
            optimizer.zero_grad()
            loss.backward()
            optimizer.step()
        
        
        out, predict_hidden = encoder(predict_en_tensor)
        out = out.to(device)
        print(zh_tensor_outputs_to_sentence(decoder.forward(predict_hidden, out)))
    print(f"Total training time: {time.time() - start_time}")


device = torch.device("mps" if torch.backends.mps.is_available() else "cpu")
encoder_trained = Encoder(embedding_dim=32, vocab_size=len(en_vocab), hidden_dim=128)
decoder_trained = LuongAttnDecoder(embedding_dim=32, vocab_size=len(zh_vocab), hidden_dim=128, device=device)
print(device)
encoder_trained.to(device)
decoder_trained.to(device)
general_attn_layer.to(device)
train(4, train_set_mini, encoder_trained, decoder_trained, device)

mps
蚀蚊哖俯鼐哖柒盈答柒
Number of trains 10000
Loss 5.936476695370674
Number of trains 20000
Loss 5.345361726331711
Number of trains 30000
Loss 5.1352989352464675
Number of trains 40000
Loss 5.0086994344949725
Number of trains 50000
Loss 4.925433911287785
Number of trains 60000
Loss 4.8515591985464095
Number of trains 70000
Loss 4.802882876157761
Number of trains 80000
Loss 4.757068572258949
Number of trains 90000
Loss 4.702026947426796
Number of trains 100000
Loss 4.675128614723683
Number of trains 110000
Loss 4.646908678662777
Number of trains 120000
Loss 4.620252806901932
Number of trains 130000
Loss 4.594933414840698
Number of trains 140000
Loss 4.578614836573601
Number of trains 150000
Loss 4.549867865610123
Number of trains 160000
Loss 4.534059087669849
Number of trains 170000
Loss 4.530960413765907
Number of trains 180000
Loss 4.508411643350124
Number of trains 190000
Loss 4.483580934113264
Number of trains 200000
Loss 4.4742356302976605
Number of trains 210000
Loss 4.457356423145533
Num

KeyboardInterrupt: 

In [ ]:
# save encoder and decoder 
torch.save(encoder_trained.state_dict(), './trained_models/baseline_encoder.pth')
torch.save(decoder_trained.state_dict(), './trained_models/baseline_decoder.pth')
torch.save(general_attn_layer.state_dict(), './trained_models/baseline_attn_layer.pth')

### ATTENTION MECHANISM
- we will now work on adding the attention mechanism
- we will try to add Luong attention 
- current reading source:   https://arxiv.org/pdf/1508.04025

### Luong attention in the model
- modfies the decoder
- at each decoding step, instead of relying only on the previous hidden state, the decoder now looks back at all encoder outputs and chooses which parts to pay attention to. 

Key ideas:
    - Encoder still processes the whole input and outputs as sequence of hidden states:

    - Decoder at time step t has its own hidden state st.
    - Attention score is computed between st (current decoder hidden state) and each hi (encoder hidden state)
    - scoring function (how you actually calculate the attention score )
        - dot product score(st, hi) = st * Wa * hi 
    
    - for all encoder outputs, softmax the scores to get a probability distribution over input positions  (adds up to 1)

    - context vector:
        - weighted sum of encoder hidden states using the attention weights 
        - a "summary" of the input that the decoder should focus on at step t 
    
    final step: combine the context vector, and the decoder hidden state to predict the next word. 

- decoder hidden --> compute scores with all encoder outputs --> softmax --> get context --> combine with decoder hidden --> predict word 


### score
- we calculate a score vector
- this score vector is softmaxed, this turns it into a probability vector (adds up to 1)
- whatever we multiply with this softmaxed score vector has the effect of being amplified or ignored. If the corresponding attention cell is close to 1, it is amplified, but if it is small, the value goes to 0

- we are going to use Luong Attention general version 

$$
score = h_t^{T} W_a h_s
$$

### preimplementation thoughts

1. Encoder stays the same. we just keep track of out, as it represents the entire hidden state across all time steps.

out, hidden = Encoder(sequence)

2. We change the teacher forcing code to be one time step at a time (not vectorized, going to map it exactly as the inference logic. 

I will apply below to both the teacher forcing and inference blocks:

3. run:
out, hidden = self.LSTM(embeds.view(1, 1, -1), hidden)

to give us the hidden state for this current time step.

3. With encoder out, representing all hidden states, and hidden representing the decoder hidden state for this current time step, we have all we need to create the SCORES tensor.
	a. We do this by passing in encoder OUT to a NEW linear layer (I still need to figure out the dimensions, but I'll just do some math during the implementation time). This is dot with decoder hidden . I know it should be out, because we want to use the ALL hidden states, that's the point of attention. 
        - remember for dot, this is matrix multiplication , so inner dimensions must match
        - the inner dimennsions for encoder out and hidden for decoder must match
        - encoder_out is [seq_len, 1, hidden_dim]
        - decoder hidden is [1, 1, hidden_dim]
        - so in other words, our output must maintain hidden_dim, so its a square matrix? lol 

    b. Now we will have a scores tensor. I softmax this so as to create a vector that can amplify / minimize whatever its multiplied by. I multiply this softmax vector with out (all encoder hidden states). This gives a weighted out tensor, I sum it up to give us a single context tensor.
        
        - since our goal is to amplify each hidden state or deamplify, based on the corresponding attention value, we want elementwise multiplication. Since they are the size of embedding dim, it should be element row multiplication. Expand the softmazed scores to match the hidden states shape (using broadcasting.)


4. with the context tensor, we concatenate with CURRENT decoder hidden state torch.cat(context, hidden)

5. We pass this combined tensor to the already existing linear layer. By already existing linear layer, i mean the second line of this code where we get the logits

                out, hidden = self.LSTM(embeds.view(1, 1, -1), hidden)
                logits = self.linear(out) # this will be self.linear(combined_tensor)
    **IMPORTANT** we now need to adjust the linear layer to be 2xhidden_dim, since our time step input tensor to the linear layer is now essentially the combinnation of 2 tensors 
6. everything else stays the same


At step 3, we need to make sure to get rid of the batch dimension 
    - use squeeze to get rid of the batch dimenision, otherwise it won't line up even with a transpose 

### Training runs
1. 04/25/25: first successful training run, 2 epochs @ 100k examples each. Model outputs non-garbage. Recognizes pronouns, and uses EOS tag. 
2. 04/25/25: Added vectorization to teacher forcing forward code. Use MPS, 3 epochs @200k examples each. Loss drops to around 4.6. Model still learns pronouns, but also common words like "need" --> 想要

3. 04/27/25: . 4 epochs at 500,000 pairs. Added Luong attention
- had to stop after one epoch because it was taking too long, but loss dropped to 4.27.  (04/28/25)
- "I love bread" --> "我爱我爱<EOS>" (Got I love)